Support for GE HealthCare (GEHC) B0 Fieldmap processing.

brice82 opened the pull request:
Correction of the existing GEHC B0 fieldmap processing (which was incomplete) and implementation of a new B0 fieldmap processing for GEHC data. The minimum required FSL-version is 6.0.7.1 for GEHC B0 fieldmap processing.

The existing fieldmap processing "GeneralElectricFieldMap" was renamed "GEHealthCareLegacyFieldMap". It is made for Legacy existing data converted with dcm2niix (pre-v1.0.20210410). These GEHC B0Maps were a 2 volumes NIFTI file (Volume 1 is the B0Map in Hertz, and Volume 2 the magnitude image).

A new fieldmap processing is introduced and named "GEHealthCareFieldMap". This is intended for new data converted with recent dcm2niix (v1.0.20220720 and later). The GEHC B0Maps are split as 2 NIfTI files (the magnitude and the b0map_fieldmaphz in Hertz - with the suffix "_fieldmaphz" for easy identification). The usage of this fieldmap processing is identical to Siemens/Philips pipeline.

For both "GEHealthCareLegacyFieldMap" and "GEHealthCareFieldMap" correction methods the "DeltaTE" is now a mandatory input.

Description of changes:
modified:   global/scripts/fsl_version.shlib
Functions to format FSL-version strings, to compare FSL-version and to check that minimun required FSL-version were added.

modified:   global/scripts/FieldMapPreprocessingAll.sh
Modification of "GeneralElectricFieldMap" renamed "GEHealthCareLegacyFieldMap" to add phase unwrapping and processing with fsl_prepare_fieldmap.
Added the "GEHealthCareFieldMap" processing via fsl_prepare_fieldmap.
The input DeltaTE is now mandatory.
And a verification that the minimum FSL-version criterion is met.

The following scripts were also modified to reflect the changes made in global/scripts/fsl_version.shlib and mainly global/scripts/FieldMapPreprocessingAll.sh.

modified:   Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatch.sh
modified:   Examples/Scripts/Historical/7T/GenericfMRIVolumeProcessingPipelineBatch.7T.sh
modified:   Examples/Scripts/PreFreeSurferPipelineBatch.sh
modified:   PreFreeSurfer/PreFreeSurferPipeline.sh
modified:   PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh
modified:   fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh
modified:   fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh

Modification tested on a database of 7 subjects containing 3DT1w, 3DT2w, resting-state fMRI, 3D-B0maps, and a pair of Spin-Echo EPI with positve and negative phase-encodings (a.k.a SpinEchoFielMaps for TOPUP).

glasserm reviewed (commented):
I would like to seem some fMRI test results.  Also, why did you modify the Philips section to remove the DeltaTE requirement?

brice82 commented:
I will sent you the fMRI data (tomorrow). I used the Topup corrected images as a reference, and the B0map corrected fMRI data are very close to the Topup corrected fMRI data. 

=> "why did you modify the Philips section to remove the DeltaTE requirement ?"
Simply because DeltaTE is now a Mandatory option in every case (Philips, Siemens, GEHC). 
So I have removed from the Philips only and Siemens only sections. 
(The DeltaTE changed from optionnal to mandatory, and it is now tested as such, so no need to retest for specific cases after. )

glasserm commented:
Okay thanks.

glasserm commented:
I have reviewed some test data from Brice and it looks good.  I am okay to merge if others are.

coalsont commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -283,7 +292,7 @@ main()
 		PhaseInputName="${StudyFolder}/${Subject}/unprocessed/3T/T1w_MPR1/${Subject}_3T_FieldMap_Phase.nii.gz"
 
-		# The TE variable should be set to 2.46ms for 3T scanner, 1.02ms for 7T
+		# The DeltaTE (echo time difference) variable should be set to 2.46ms for 3T scanner, 1.02ms for 7T
```
If we wanted to, we could clarify that these example numbers are for HCP YA data.

brice82 replied:
Yes, it sound like a good opportunity

brice82 replied:
TE was repalced by DeltaTE in PreFreeSurferPipelineBatch.sh

coalsont commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -208,11 +211,11 @@ opts_AddMandatory '--brainsize' 'BrainSize' 'size_value' "Brain size estimate in
+opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'file_path' "Siemens/Philips/GE HealthCare Gradient Echo Fieldmap magnitude file"
 
-opts_AddOptional '--fmapphase' 'PhaseInputName' 'file_path' "Siemens/Philips Gradient Echo Fieldmap phase file"
+opts_AddOptional '--fmapphase' 'PhaseInputName' 'file_path' "Siemens/Philips Gradient Echo Fieldmap phase file or GE HealthCare Fielmap in Hertz"
```
```suggestion
opts_AddOptional '--fmapphase' 'PhaseInputName' 'file_path' "Siemens/Philips Gradient Echo Fieldmap phase file or GE HealthCare Fieldmap in Hertz"
```

coalsont commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -319,6 +298,27 @@ case "$DistortionCorrection" in
+		fi
+    # Check that FSL is at least the minimum required FSL version, abort if needed (and log FSL-version)
+    fsl_minimum_required_version_check "$GEHEALTHCARE_MINIMUM_FSL_VERSION" \
+      "For ${DistortionCorrection} method the minimun required FSL version is ${GEHEALTHCARE_MINIMUM_FSL_VERSION}. "
```
```suggestion
      "For ${DistortionCorrection} method the minimum required FSL version is ${GEHEALTHCARE_MINIMUM_FSL_VERSION}. "
```

coalsont commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -64,9 +67,13 @@ opts_AddMandatory '--method' 'DistortionCorrection' 'method' "method used for re
+           This option is maintained for backward compatibility.
+
+        '${GE_HEALTHCARE_METHOD_OPT}'
+             use GE HealthCare specific Gradient Echo Field Maps for SDC (fieldmap in Hz and magnitude in two separate NIfTI files via --fmapphase and --fmapmag).
```
Indentation here doesn't match the other help info.  Same comment applies to the help info parts in other scripts, but the expected indentation is also different than here...

coalsont replied:
```suggestion
           use GE HealthCare specific Gradient Echo Field Maps for SDC (fieldmap in Hz and magnitude in two separate NIfTI files via --fmapphase and --fmapmag).
```

coalsont commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -79,7 +85,11 @@ opts_AddMandatory '--dcmethod' 'DistortionCorrection' 'method' "Which method to
 
         '${GENERAL_ELECTRIC_METHOD_OPT}'
-             use General Electric specific Gradient Echo Field Maps for SDC
+           use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --fmapgeneralelectric argument).
```
Indentation, per above

coalsont commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -82,7 +85,11 @@ opts_AddMandatory '--method' 'DistortionCorrection' 'method' "method to use for
 
         '${GENERAL_ELECTRIC_METHOD_OPT}'
-             use General Electric specific Gradient Echo Field Maps for SDC
+           use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --fmapgeneralelectric argument).
```
Indentation

coalsont commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -82,7 +85,11 @@ opts_AddMandatory '--method' 'DistortionCorrection' 'method' "method to use for
 
         '${GENERAL_ELECTRIC_METHOD_OPT}'
-             use General Electric specific Gradient Echo Field Maps for SDC
+           use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --fmapgeneralelectric argument).
+           This option is maintained for backward compatibility.
```
```suggestion
             use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --fmapgeneralelectric argument).
             This option is maintained for backward compatibility.
```

coalsont commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -79,7 +85,11 @@ opts_AddMandatory '--dcmethod' 'DistortionCorrection' 'method' "Which method to
 
         '${GENERAL_ELECTRIC_METHOD_OPT}'
-             use General Electric specific Gradient Echo Field Maps for SDC
+           use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --fmapgeneralelectric argument).
+           This option is maintained for backward compatibility.
```
```suggestion
             use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --fmapgeneralelectric argument).
             This option is maintained for backward compatibility.
```

coalsont commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -141,18 +180,29 @@ case $DistortionCorrection in
-		    mv vol0000.nii.gz ${WD}/FieldMap_deg.nii.gz
-		    mv vol0001.nii.gz ${WD}/Magnitude.nii.gz
+            ${FSLDIR}/bin/fslsplit ${GEB0InputName}     # split image into vol0000=fieldmap and vol0001=magnitude
+		    ${FSLDIR}/bin/immv vol0000 ${WD}/FieldMapHertz
+		    ${FSLDIR}/bin/immv vol0001 ${WD}/Magnitude
 		    ${FSLDIR}/bin/bet ${WD}/Magnitude ${WD}/Magnitude_brain -f 0.35 -m #Brain extract the magnitude image
-		    ${FSLDIR}/bin/fslmaths ${WD}/FieldMap_deg.nii.gz -mul 6.28 ${WD}/FieldMap.nii.gz
+		    ${FSLDIR}/bin/fsl_prepare_fieldmap GEHC_FIELDMAPHZ ${WD}/FieldMapHertz ${WD}/Magnitude_brain ${WD}/FieldMap ${DeltaTE}
```
This has some tabs vs spaces going on, and since you are editing these lines anyway (so git blame will show this PR), can you clean that up?  Personally, I prefer 4 spaces per level, but we don't really have a consensus (other than "consecutive lines should use the same convention").

brice82 replied:
Not sure what happen here. I normally use 4 space as well.

coalsont commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -141,18 +180,29 @@ case $DistortionCorrection in
+        # -- GE HealthCare Gradient Echo Field Maps --
+        # -------------------------------------------- 
+        	
+            ${FSLDIR}/bin/fslmaths ${MagnitudeInputName} -Tmean ${WD}/Magnitude #normally only one volume 
+		    ${FSLDIR}/bin/bet ${WD}/Magnitude ${WD}/Magnitude_brain -f 0.35 -m #Brain extract the magnitude image
+		    ${FSLDIR}/bin/imcp ${PhaseInputName} ${WD}/FieldMapHertz
+		    ${FSLDIR}/bin/fsl_prepare_fieldmap GEHC_FIELDMAPHZ ${WD}/FieldMapHertz ${WD}/Magnitude_brain ${WD}/FieldMap ${DeltaTE}
```
```suggestion
        ${FSLDIR}/bin/fslmaths ${MagnitudeInputName} -Tmean ${WD}/Magnitude #normally only one volume 
        ${FSLDIR}/bin/bet ${WD}/Magnitude ${WD}/Magnitude_brain -f 0.35 -m #Brain extract the magnitude image
        ${FSLDIR}/bin/imcp ${PhaseInputName} ${WD}/FieldMapHertz
        ${FSLDIR}/bin/fsl_prepare_fieldmap GEHC_FIELDMAPHZ ${WD}/FieldMapHertz ${WD}/Magnitude_brain ${WD}/FieldMap ${DeltaTE}
```

coalsont commented on global/scripts/fsl_version.shlib:
```diff
@@ -10,20 +16,162 @@
+#  in argument 2 ($2).  
+#
+# Usage Example:
+#  fsl_minimum_required_version_check "6.0.7.1" "My Optionnal Error Message"
```
```suggestion
#  fsl_minimum_required_version_check "6.0.7.1" "My Optional Error Message"
```

coalsont commented on global/scripts/fsl_version.shlib:
```diff
@@ -10,20 +16,162 @@
+    min_criterion_met=$(fsl_minimum_required_version "$min_ver")
+
+    if [ "$min_criterion_met" == "true" ]; then
+        log_Msg "INFO: The minimum FSL version criterion is met (mininum-version: ${min_ver} <= current-version ${cur_ver})"
+    else
+        if [[ -n "$my_err_msg" ]]; then
+            log_Err "$my_err_msg"
+        fi
+        log_Err_Abort "The minimum FSL version criterion is NOT met (mininum-version: ${min_ver} > current-version ${cur_ver})"
```
```suggestion
        log_Msg "INFO: The minimum FSL version criterion is met (minimum-version: ${min_ver} <= current-version ${cur_ver})"
    else
        if [[ -n "$my_err_msg" ]]; then
            log_Err "$my_err_msg"
        fi
        log_Err_Abort "The minimum FSL version criterion is NOT met (minimum-version: ${min_ver} > current-version ${cur_ver})"
```

coalsont commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -74,32 +91,54 @@ case $DistortionCorrection in
+        
+        # Check that FSL is at least the minimum required FSL version, abort if needed (and log FSL-version)
+        fsl_minimum_required_version_check "$GEHEALTHCARE_MINIMUM_FSL_VERSION" \
+            "For $DistortionCorrection method the minimun required FSL version is ${GEHEALTHCARE_MINIMUM_FSL_VERSION}."
```
```suggestion
            "For $DistortionCorrection method the minimum required FSL version is ${GEHEALTHCARE_MINIMUM_FSL_VERSION}."
```

coalsont commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -319,6 +298,27 @@ case "$DistortionCorrection" in
+    fi
+    # Check that FSL is at least the minimum required FSL version, abort if needed (and log FSL-version)
+    fsl_minimum_required_version_check "$GEHEALTHCARE_MINIMUM_FSL_VERSION" \
+      "For ${DistortionCorrection} method the minimun required FSL version is ${GEHEALTHCARE_MINIMUM_FSL_VERSION}. " 
```
```suggestion
      "For ${DistortionCorrection} method the minimum required FSL version is ${GEHEALTHCARE_MINIMUM_FSL_VERSION}. " 
```

coalsont commented on global/scripts/fsl_version.shlib:
```diff
@@ -10,20 +16,162 @@
+# Usage Example:
+#  ret_val=$(fsl_minimum_required_version "6.0.7.2")
+#  if [ "$ret_val" == "false" ]; then
+#      log_Msg "ERROR: FSL minimun required version is 6.0.7.2. 
```
```suggestion
#      log_Msg "ERROR: FSL minimum required version is 6.0.7.2. 
```

coalsont commented:
Logic looks reasonable.  I have made some documentation/formatting/typo suggestions.

coalsont commented on global/scripts/fsl_version.shlib:
```diff
@@ -10,20 +16,162 @@
+        return 0
+    fi
+
+    # split version by dots - assuming the format X.Y.Z.W
+    for k in $(seq 0 $max_ndots); 
+    do 
+        cur_ver_array[$k]=$(awk -F"." -v i=$((k+1)) '{print $i}' <<< "$cur_ver")
+        min_ver_array[$k]=$(awk -F"." -v i=$((k+1)) '{print $i}' <<< "$min_ver")
+    done
```
```suggestion
    IFS=. read -a cur_ver_array <<< "$cur_ver"
    IFS=. read -a min_ver_array <<< "$min_ver"
```

You could consider using the array lengths in the test loop to handle the versions as-is, rather than preformatting.

brice82 replied:
In this case, we also need to consider arbitrary length and find the maximum number of dots from `$cur_ver`and `$min_ver` for the subsequent comparison (otherwise it won't work with more that 3 dots). This will make things much more generic (and the comparison via fsl_minimum_required_version function will work any FSL-version string length).  
Great suggestions. 
I did that and the test results are: 
```
==> Test variable and unsual FSL string length
==> Test fsl_version_format on unsual FSL string length
Test fsl_version_format 6 return 6.0.0.0
Test fsl_version_format 7.1.2.0.1 return 7.1.2.0.1
Test fsl_version_format 6.0.7.6.0 return 6.0.7.6.0
Test fsl_version_format 6.0.7.6.1 return 6.0.7.6.1
Test fsl_version_format 6.0.7.6.0.0 return 6.0.7.6.0.0
Test fsl_version_format 6.0.7.6.0.1 return 6.0.7.6.0.1
====
current FSL-Version = 6.0.7.6
==> Test function fsl_minimum_required_version on unsual FSL string length: 
Test: minimum-ver <= current-ver is ? 
Test:               6 <= 6.0.7.6 is true
Test:       7.1.2.0.1 <= 6.0.7.6 is false
Test:       6.0.7.6.0 <= 6.0.7.6 is true
Test:       6.0.7.6.1 <= 6.0.7.6 is false
Test:     6.0.7.6.0.0 <= 6.0.7.6 is true
Test:     6.0.7.6.0.1 <= 6.0.7.6 is false 
```
I upload the modify version of fsl_version.shlib to resolve the issue.

brice82 replied:
I Uploaded the modification for arbitrary length of FSL-version string. 
The consequence, is the `fsl_version_format()` and the constant `MAX_NDOTS_IN_FSLVERSION=3` are just for cosmetic. I left these functions.
Do you want to remove them ?

brice82 commented:
Quick summary of changes for the today's modification.
Following Tim suggestions the function `fsl_minimum_required_version()` is able to compare FSL-version string of arbitrary length.

- I left function `fsl_version_format()` and constant `MAX_NDOTS_IN_FSLVERSION=3` for cosmetic purpose. Just let me know if you want me to remove them. 
- I fixed the indentation in `FieldMapProcessingAll.sh`. 
- Thanks Tim for the Typos corrections.

coalsont commented on global/scripts/fsl_version.shlib:
```diff
@@ -10,20 +16,170 @@
+    IFS=. read -a cur_ver_array <<< "$cur_ver"
+    IFS=. read -a min_ver_array <<< "$min_ver"
+
+    # count the dots
+    ndots_cur_ver=$(awk -F"." '{print NF-1}' <<< "$cur_ver")
+    ndots_min_ver=$(awk -F"." '{print NF-1}' <<< "$min_ver")
```
```suggestion
    # count the dots
    ndots_cur_ver=$((${#cur_ver_array[@]} - 1))
    ndots_min_ver=$((${#min_ver_array[@]} - 1))
```
Or if you feel like it, rework the code a bit to use the array lengths rather than number of separators.

brice82 replied:
This makes a lot of sense.

brice82 replied:
integrated

coalsont commented on global/scripts/fsl_version.shlib:
```diff
@@ -10,20 +16,170 @@
+        max_ndots=$ndots_min_ver
+    fi
+
+    for k in $(seq 0 $max_ndots)
```
You should be able to stop at the minimum number of dots instead of the maximum, if you don't treat `3.0.0` as being equal to `3.0`.

coalsont replied:
dpkg implements "x.0 is newer than x", and while it has operators that are described as "treat empty version as later", that doesn't change these cases, they literally mean "if the whole version string is the empty string, it is greater than any nonempty string":
```
$ if dpkg --compare-versions 3.0 'lt-nl' 3.0.0; then echo yes; fi
yes
```

I suggest this, which doesn't need to add zeros:
```bash
min_fields=$((${#cur_ver_array[@]}))
if ((${#req_ver_array[@]} < min_fields))
then
    min_fields=$((${#req_ver_array[@]}))
fi
for ((i = 0; i < min_fields; ++i))
    if ((cur_ver_array[i] < req_ver_array[i])); then echo false; return; fi
    if ((cur_ver_array[i] > req_ver_array[i])); then echo true; return; fi
end
if ((${#cur_ver_array[@]} >= ${#req_ver_array[@]})); then echo true; return; fi
echo false; return
```

brice82 replied:
Right, I have treated "x.0 as equal to x" but I am also fine with "x.0 is newer than x". 
I'll try and rerun test.

brice82 replied:
The comparison is now "x.0 is newer than x".
So before modification it was returning
```
Test:       6.0.7.6.0 <= 6.0.7.6 is true
Test:     6.0.7.6.0.0 <= 6.0.7.6 is true
```
and now these cases are false: 
```
Test:       6.0.7.6.0 <= 6.0.7.6 is false
Test:     6.0.7.6.0.0 <= 6.0.7.6 is false
``` 
All other cases returns the same results. 
I attached test results and and test script. 

[fsl_version_test_before.txt](https://github.com/Washington-University/HCPpipelines/files/14319071/fsl_version_test_before.txt)
[fsl_version_test_after.txt](https://github.com/Washington-University/HCPpipelines/files/14319072/fsl_version_test_after.txt)
[test_modification_fslversion_functions.sh.txt](https://github.com/Washington-University/HCPpipelines/files/14319076/test_modification_fslversion_functions.sh.txt)

coalsont commented on Examples/Scripts/Historical/7T/GenericfMRIVolumeProcessingPipelineBatch.7T.sh:
```diff
@@ -208,11 +208,20 @@ do
+		# For GE HealthCare, see related notes in PreFreeSurferPipelineBatch.sh and FieldMapProcessingAll.sh
+
+		# Path to GE HEalthCare Legacy style B0 fieldmap with two volumes
+        #   1. field map in hertz
+        #   2. magnitude
+        # Set to "NONE" if not using "GEHealthCareLegacyFieldMap" as the value for the DistortionCorrection variable
+        #
+        # Example Value: 
+        #  GEB0InputName="${StudyFolder}/${Subject}/unprocessed/3T/${fMRIName}/${Subject}_3T_GradientEchoFieldMap.nii.gz" 
+        #  DeltaTE=2.272 # ms 
```
These indent with spaces, surrounding lines are tabs.

coalsont replied:
```suggestion
		#   1. field map in hertz
		#   2. magnitude
		# Set to "NONE" if not using "GEHealthCareLegacyFieldMap" as the value for the DistortionCorrection variable
		#
		 # Example Value: 
		#  GEB0InputName="${StudyFolder}/${Subject}/unprocessed/3T/${fMRIName}/${Subject}_3T_GradientEchoFieldMap.nii.gz" 
		#  DeltaTE=2.272 # ms 
```

coalsont commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -283,7 +292,7 @@ main()
 		PhaseInputName="${StudyFolder}/${Subject}/unprocessed/3T/T1w_MPR1/${Subject}_3T_FieldMap_Phase.nii.gz"
 
-		# The TE variable should be set to 2.46ms for 3T scanner, 1.02ms for 7T
+		# The DeltaTE (echo time difference) variable should be set to 2.46ms for 3T scanner, 1.02ms for 7T
 		# scanner or "NONE" if not using
 		TE="2.46"
```
```suggestion
		# The DeltaTE (echo time difference) of the fieldmap.  For HCP Young Adult data, this variable should be set to 2.46ms for 3T scans, 1.02ms for 7T
		# scans, or "NONE" if not using readout distortion correction
		TE="2.46"
```

brice82 commented:
Following @coalsont suggestions, the FSL-version string comparison consider "x.0 is newer than x" instead of "x.0 is equal to x".

mharms commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -252,9 +252,18 @@ main()
-		#   "GeneralElectricFieldMap"
-		#     Average any repeats and use General Electric specific Gradient
-		#     Echo Field Map for readout distortion correction
+		#   "GEHealthCareLegacyFieldMap"
```
Are we removing "GeneralElectricFieldMap" entirely as an option?  Is there any reason that we can't/shouldn't continue to support that as an alternative option name for backwards compatibility?

glasserm replied:
I think it was left in the actual parsing code.

mharms replied:
Maybe I missed it, but I don't think so...

mharms replied:
Ok.  Given the need to now supply `DeltaTE`, it seems reasonable to no longer support the old `GeneralElectricFieldMap` as an option, and good that we renamed it (as you just did).

mharms commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatch.sh:
```diff
@@ -227,17 +227,20 @@ for Subject in $Subjlist ; do
-        # Path to General Electric style B0 fieldmap with two volumes
-        #   1. field map in degrees
+        # or set the following inputs if using regular FIELDMAP (i.e. SiemensFieldMap GEHealthCareFieldMap PhilipsFieldMap)
+        MagnitudeInputName="NONE" #Expects 4D Magnitude volume with two 3D volumes (differing echo times) - or a single 3D Volume
```
This may be confusing to users.  If supplying just a single 3D Volume, does it matter whether it is the magnitude image of the 1st or 2nd echo?

brice82 replied:
GEHC outputs only the magnitude image of the echo 1. However, I kept the same philosophy as for siemens/philipps and if a 4D magnitude image if supplied, then it is averaged. 
I have added the "- or a single 3D Volume" to stress that the 4D magnitude was not mandatory. 

I understood that Siemens outputs the 2 magnitude images but not 100% sure.

mharms commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatch.sh:
```diff
@@ -227,17 +227,20 @@ for Subject in $Subjlist ; do
+        # or set the following inputs if using regular FIELDMAP (i.e. SiemensFieldMap GEHealthCareFieldMap PhilipsFieldMap)
+        MagnitudeInputName="NONE" #Expects 4D Magnitude volume with two 3D volumes (differing echo times) - or a single 3D Volume
+        PhaseInputName="NONE" #Expects a 3D Phase difference volume (Siemen's style) -or Fieldmap in Hertz for GE Healthcare
+        DeltaTE="NONE" #2.46ms for 3T, 1.02ms for 7T - for GE Healthcare at 3.0T, *usually* 2.304ms for 2D-B0MAP and 2.272 ms 3D B0MAP.
```
```suggestion
        DeltaTE="NONE" #For Siemens, typically 2.46ms for 3T, 1.02ms for 7T; For GE Healthcare at 3T, *usually* 2.304ms for 2D-B0MAP and 2.272 ms 3D B0MAP
```

mharms commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatch.sh:
```diff
@@ -227,17 +227,20 @@ for Subject in $Subjlist ; do
+        DeltaTE="NONE" #2.46ms for 3T, 1.02ms for 7T - for GE Healthcare at 3.0T, *usually* 2.304ms for 2D-B0MAP and 2.272 ms 3D B0MAP.
+        # For GE HealthCare, see related notes in PreFreeSurferPipelineBatch.sh and FieldMapProcessingAll.sh
+
+        # Path to GE HEalthCare Legacy style B0 fieldmap with two volumes
```
```suggestion
        # Path to GE HealthCare Legacy style B0 fieldmap with two volumes
```

mharms commented on Examples/Scripts/GenericfMRIVolumeProcessingPipelineBatch.sh:
```diff
@@ -227,17 +227,20 @@ for Subject in $Subjlist ; do
+
+        # Path to GE HEalthCare Legacy style B0 fieldmap with two volumes
+        #   1. field map in hertz
         #   2. magnitude
```
```suggestion
        #   2. magnitude image
```

mharms commented on Examples/Scripts/Historical/7T/GenericfMRIVolumeProcessingPipelineBatch.7T.sh:
```diff
@@ -208,11 +208,20 @@ do
-		DeltaTE="NONE" #2.46ms for 3T, 1.02ms for 7T
-		
-		# Not using General Electric Gradient Echo Field Maps for Distortion Correction
+		# or set the following inputs if using regular FIELDMAP (i.e. SiemensFieldMap GEHealthCareFieldMap PhilipsFieldMap)
```
Implement same changes to the ".7T.sh" script that I suggested for the "3T" version.

mharms commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -252,9 +252,18 @@ main()
+		#   "GEHealthCareLegacyFieldMap"
+		#     Average any repeats and use GE HeathCare Legacy specific Gradient
+		#     Echo Field Map for readout distortion correction. 
+		#	  The Legacy fieldmap is a two volume file: 1. field map in Hz and 2. magnitude.
```
```suggestion
		#	  The Legacy fieldmap is a two volume file: 1. field map in Hz and 2. magnitude image.
```

mharms commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -252,9 +252,18 @@ main()
-		#     Average any repeats and use General Electric specific Gradient
-		#     Echo Field Map for readout distortion correction
+		#   "GEHealthCareLegacyFieldMap"
+		#     Average any repeats and use GE HeathCare Legacy specific Gradient
```
Repeats of what?

brice82 replied:
This is the averaging of any repeats of the magnitude. 
Such as we have consistent behavior and possibilities for all manfacturer. 
(I has also reused the exact same wording)

mharms commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -252,9 +252,18 @@ main()
+		#	"GEHealthCareFieldMap" 
+		#	  Average any repeats and use GE HealthCare specific Gradient Echo
+		#     Field Maps for readout distortion correction
+		#	  This uses two separate NIfTI file for the fieldmap in Hz and the magnitude
```
```suggestion
		#	  This uses separate NIfTI files for the fieldmap in Hz and the magnitude image
```

mharms commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -252,9 +252,18 @@ main()
+		#     Echo Field Map for readout distortion correction. 
+		#	  The Legacy fieldmap is a two volume file: 1. field map in Hz and 2. magnitude.
+		# 	  Use "GEB0InputName" variable to specify the 2-Volume file. 
+		#	  Set "TE" variable to the EchoTime difference (TE2-TE1). 
```
This could be confusing.  If "TE" is really supposed to "DeltaTE", why not rename the variable to "DeltaTE" instead.

brice82 replied:
I'll fix that

mharms commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -283,8 +292,8 @@ main()
 
-		# The TE variable should be set to 2.46ms for 3T scanner, 1.02ms for 7T
-		# scanner or "NONE" if not using
+		# The DeltaTE (echo time difference) of the fieldmap.  For HCP Young Adult data, this variable should be set to 2.46ms for 3T scans, 1.02ms for 7T
```
```suggestion
		# The DeltaTE (echo time difference) of the fieldmap.  For HCP Young Adult data, this variable would typically be 2.46ms for 3T scans, 1.02ms for 7T
```

mharms commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -346,25 +355,51 @@ main()
 		# volumes
-		#   1) fieldmap in deg and
+		#   1) fieldmap in hertz and
 		#   2) magnitude,
```
```suggestion
		#   2) magnitude image,
```

mharms commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -346,25 +355,51 @@ main()
 		#   GEB0InputName="${StudyFolder}/${Subject}/unprocessed/3T/T1w_MPR1/${Subject}_3T_GradientEchoFieldMap.nii.gz"
+		#	TE=2.304 
+		#   Here TE refers to the DeltaTE in ms
+		#   NOTE: At 3.0T, the DeltaTE is *usually* 2.304ms for 2D-B0MAP and 2.272 ms 3D B0MAP.
```
```suggestion
		#   NOTE: At 3T, the DeltaTE is *usually* 2.304ms for 2D-B0MAP and 2.272ms 3D-B0MAP.
```

mharms commented on Examples/Scripts/PreFreeSurferPipelineBatch.sh:
```diff
@@ -346,25 +355,51 @@ main()
+		#   NOTE: At 3.0T, the DeltaTE is *usually* 2.304ms for 2D-B0MAP and 2.272 ms 3D B0MAP.
+		#   NOTE: The Delta can be found in json files (if data converted with recent dcm2niix)
+		#   NOTE: Then DeltaTE = (EchoTime2-EchoTime1)*1000
+		#	NOTE: In the DICOM DeltaTE = round(abs(1e6/( 2*pi*(0019,10E2) )))
```
```suggestion
		#	NOTE: In the DICOM, DeltaTE = round(abs(1e6/( 2*pi*(0019,10E2) )))
```

mharms commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -208,11 +211,11 @@ opts_AddMandatory '--brainsize' 'BrainSize' 'size_value' "Brain size estimate in
+opts_AddOptional '--fmapphase' 'PhaseInputName' 'file_path' "Siemens/Philips Gradient Echo Fieldmap phase file or GE HealthCare Fieldmap in Hertz"
 
-opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'file_path' "General Electric Gradient Echo Field Map file  Two volumes in one file  1. field map in deg  2. magnitude"
+opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'file_path' "GE HealthCare Legacy Gradient Echo Field Map file. Two volumes in one file  1. field map in hertz  2. magnitude"
```
```suggestion
opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'file_path' "GE HealthCare Legacy Gradient Echo Field Map approach, which contains two volumes in one file:  1. field map in hertz;  2. magnitude image"
```

mharms replied:
Also, should we perhaps change this flag to `--fmapgelegacy` as the forward-facing name, with `--fmapgeneralelectric` still allowed as an alternative? (I think the library allows specification of alternative argument flags, right @coalsont ?)

brice82 replied:
No problem to change `--fmapgeneralelectric` in `--fmapgelegacy`. 
Having the 2 possibilities might be confusing, no ? and complicate maintenance.

mharms replied:
I guess since the old approach no longer even exists conceptually, it makes sense to drop `--fmapgeneralelectric` as a flag entirely.  Anyone trying to use it (e.g., with existing scripts) will get any error anyway that `DeltaTE` doesn't exist.   @coalsont: do you agree?

coalsont replied:
Yes, providing one or more legacy flags after the default value (quoted empty string if you want the same behavior as no default value) will take it as another way to do the same thing (but it is not mentioned in the help info).  If we don't want to accept `--fmap`, then there is no need for a legacy flag here.  I think we can/should take `--fmap` here, and, per below, if we wanted to change the name at all, the flag should be brand-agnostic, such as `--fmapcombined`, even if only one brand used it so far.

I don't think any of the flags should contain brand names, they should simply be used for handling different arrangements of the input data.

One of the two new methods does take data arranged in the `--fmap` format.

mharms commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -244,10 +247,13 @@ opts_AddMandatory '--avgrdcmethod' 'AvgrdcSTRING' 'avgrdcmethod' "Averaging and
-      average any repeats and use General Electric specific Gradient
-      Echo Field Maps for readout distortion correction
+  '${GENERAL_ELECTRIC_METHOD_OPT}'
+      use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --fmapgeneralelectric argument).
```
```suggestion
      use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (i.e., field map in Hz and magnitude image in a single NIfTI file, via --fmapgeneralelectric argument).
```

mharms commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -244,10 +247,13 @@ opts_AddMandatory '--avgrdcmethod' 'AvgrdcSTRING' 'avgrdcmethod' "Averaging and
+      This option is maintained for backward compatibility.
 
+  '${GE_HEALTHCARE_METHOD_OPT}'
+      use GE HealthCare specific Gradient Echo Field Maps for SDC (fieldmap in Hz and magnitude in two separate NIfTI files via --fmapphase and --fmapmag).
```
```suggestion
      use GE HealthCare specific Gradient Echo Field Maps for SDC (i.e., field map in Hz and magnitude image in two separate NIfTI files, via --fmapphase and --fmapmag).
```

mharms commented on PreFreeSurfer/PreFreeSurferPipeline.sh:
```diff
@@ -152,7 +152,10 @@ NONE_METHOD_OPT="NONE"
-GENERAL_ELECTRIC_METHOD_OPT="GeneralElectricFieldMap"
+# For GE HealthCare Fieldmap Distortion Correction methods 
+# see explanations in global/scripts/FieldMapPreprocessingAll.sh
+GENERAL_ELECTRIC_METHOD_OPT="GEHealthCareLegacyFieldMap" 
```
Should we perhaps change this to `GE_LEGACY_METHOD_OPT`? (Here and throughout code?)

brice82 replied:
Not a problem

mharms commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -64,9 +67,13 @@ opts_AddMandatory '--method' 'DistortionCorrection' 'method' "method used for re
+        
         '${GENERAL_ELECTRIC_METHOD_OPT}'
-           use General Electric specific Gradient Echo Field Maps for readout distortion correction
+           use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --fmapgeneralelectric argument).
```
```suggestion
           use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (i.e., fieldmap in Hz and magnitude image in a single NIfTI file, via --fmapgeneralelectric argument).
```

mharms commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -64,9 +67,13 @@ opts_AddMandatory '--method' 'DistortionCorrection' 'method' "method used for re
+           This option is maintained for backward compatibility.
+
+        '${GE_HEALTHCARE_METHOD_OPT}'
+           use GE HealthCare specific Gradient Echo Field Maps for SDC (fieldmap in Hz and magnitude in two separate NIfTI files via --fmapphase and --fmapmag).
```
```suggestion
           use GE HealthCare specific Gradient Echo Field Maps for SDC (i.e., field map in Hz and magnitude image in two separate NIfTI files, via --fmapphase and --fmapmag).
```

mharms commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -81,9 +88,9 @@ opts_AddOptional '--workingdir' 'WD' 'path' "working directory" "."
 opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'image' "input fieldmap magnitude image"
 
-opts_AddOptional '--fmapphase' 'PhaseInputName' 'image' "input fieldmap phase images (single 4D image containing 2x3D volumes)"
+opts_AddOptional '--fmapphase' 'PhaseInputName' 'image' "input fieldmap phase images in radians (Siemens/Philips) or in hertz (GE HealthCare)"
```
```suggestion
opts_AddOptional '--fmapphase' 'PhaseInputName' 'image' "input fieldmap phase images in radians (Siemens/Philips) or in Hz (GE HealthCare)"
```

mharms commented on PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh:
```diff
@@ -81,9 +88,9 @@ opts_AddOptional '--workingdir' 'WD' 'path' "working directory" "."
+opts_AddOptional '--fmapphase' 'PhaseInputName' 'image' "input fieldmap phase images in radians (Siemens/Philips) or in hertz (GE HealthCare)"
 
-opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'deg OR magnitude' "input General Electric field map (two volumes: 1. field map in deg or  2. magnitude)"
+opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'image' "input GE HealthCare Legacy field map only (two volumes: 1. field map in Hz or  2. magnitude)"
```
```suggestion
opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'image' "input GE HealthCare Legacy field map only (two volumes: 1. field map in Hz and 2. magnitude image)"
```

mharms replied:
Also, change to `--fmapgelegacy` per my comment in the other script?

mharms commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -79,7 +85,11 @@ opts_AddMandatory '--dcmethod' 'DistortionCorrection' 'method' "Which method to
 
         '${GENERAL_ELECTRIC_METHOD_OPT}'
-             use General Electric specific Gradient Echo Field Maps for SDC
+             use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --fmapgeneralelectric argument).
```
```suggestion
             use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (field map in Hz and magnitude iimage n a single NIfTI file via, --fmapgeneralelectric argument).
```

mharms commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -79,7 +85,11 @@ opts_AddMandatory '--dcmethod' 'DistortionCorrection' 'method' "Which method to
+             This option is maintained for backward compatibility.
+
+        '${GE_HEALTHCARE_METHOD_OPT}'
+             use GE HealthCare specific Gradient Echo Field Maps for SDC (fieldmap in Hz and magnitude in two separate NIfTI files via --fmapphase and --fmapmag).
```
```suggestion
             use GE HealthCare specific Gradient Echo Field Maps for SDC (field map in Hz and magnitude image in two separate NIfTI files, via --fmapphase and --fmapmag).
```

mharms commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -98,13 +108,13 @@ opts_AddOptional '--SEPhasePos' 'SpinEchoPhaseEncodePositive' 'file' "positive p
 opts_AddOptional '--topupconfig' 'TopupConfig' 'file' "Which topup config file to use"
 
-opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'file' "Siemens field map magnitude image to use"
+opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'file' "input field map magnitude image"
```
```suggestion
opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'file' "field map magnitude image"
```

mharms commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -98,13 +108,13 @@ opts_AddOptional '--SEPhasePos' 'SpinEchoPhaseEncodePositive' 'file' "positive p
+opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'file' "input field map magnitude image"
 
-opts_AddOptional '--fmapphase' 'PhaseInputName' 'file' "Siemens field map phase image to use"
+opts_AddOptional '--fmapphase' 'PhaseInputName' 'file' "input fieldmap phase images in radians (Siemens/Philips) or in hertz (GE HealthCare)"
```
```suggestion
opts_AddOptional '--fmapphase' 'PhaseInputName' 'file' "fieldmap phase images in radians (Siemens/Philips) or in Hz (GE HealthCare)"
```

mharms commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -98,13 +108,13 @@ opts_AddOptional '--SEPhasePos' 'SpinEchoPhaseEncodePositive' 'file' "positive p
 opts_AddOptional '--echodiff' 'deltaTE' 'milliseconds' "Difference of echo times for fieldmap, in milliseconds"
 
-opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'file' "General Electric field map image to use"
+opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'file' "input GE HealthCare Legacy field map only (two volumes: 1. field map in Hz or  2. magnitude)"
```
```suggestion
opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'file' "GE HealthCare Legacy field map only (two volumes: 1. field map in Hz and 2. magnitude image)"
```

mharms commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -82,7 +85,11 @@ opts_AddMandatory '--method' 'DistortionCorrection' 'method' "method to use for
 
         '${GENERAL_ELECTRIC_METHOD_OPT}'
-             use General Electric specific Gradient Echo Field Maps for SDC
+             use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --fmapgeneralelectric argument).
```
```suggestion
             use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (field map in Hz and magnitude image in a single NIfTI file, via --fmapgeneralelectric argument).
```

mharms commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -82,7 +85,11 @@ opts_AddMandatory '--method' 'DistortionCorrection' 'method' "method to use for
+             This option is maintained for backward compatibility.
+
+        '${GE_HEALTHCARE_METHOD_OPT}'
+             use GE HealthCare specific Gradient Echo Field Maps for SDC (fieldmap in Hz and magnitude in two separate NIfTI files via --fmapphase and --fmapmag).
```
```suggestion
             use GE HealthCare specific Gradient Echo Field Maps for SDC (field map in Hz and magnitude image in two separate NIfTI files, via --fmapphase and --fmapmag).
```

mharms commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -106,13 +113,13 @@ opts_AddOptional '--topupconfig' 'TopupConfig' 'file' "topup config file"
 opts_AddOptional '--subjectfolder' 'SubjectFolder' 'path' "subject processing folder"
 
-opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'field_map' "input Siemens field map magnitude image"
+opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'field_map' "input field map magnitude image"
```
```suggestion
opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'field_map' "field map magnitude image"
```

mharms commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -106,13 +113,13 @@ opts_AddOptional '--topupconfig' 'TopupConfig' 'file' "topup config file"
+opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'field_map' "input field map magnitude image"
 
-opts_AddOptional '--fmapphase' 'PhaseInputName' 'image' "input Siemens field map phase image"
+opts_AddOptional '--fmapphase' 'PhaseInputName' 'image' "input fieldmap phase images in radians (Siemens/Philips) or in hertz (GE HealthCare)"
```
```suggestion
opts_AddOptional '--fmapphase' 'PhaseInputName' 'image' "fieldmap phase images in radians (Siemens/Philips) or in Hz (GE HealthCare)"
```

mharms commented on fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh:
```diff
@@ -106,13 +113,13 @@ opts_AddOptional '--topupconfig' 'TopupConfig' 'file' "topup config file"
 opts_AddOptional '--echodiff' 'deltaTE' 'number (milliseconds)' "difference of echo times for fieldmap, in milliseconds"
 
-opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'image' "input General Electric field map image"
+opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'image' "input GE HealthCare Legacy field map only (two volumes: 1. field map in Hz or  2. magnitude)"
```
```suggestion
opts_AddOptional '--fmapgeneralelectric' 'GEB0InputName' 'image' "GE HealthCare Legacy field map only (two volumes: 1. field map in Hz  and 2. magnitude image)"
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -9,9 +9,23 @@
+GENERAL_ELECTRIC_METHOD_OPT="GEHealthCareLegacyFieldMap"
+# "GEHealthCareLegacyFieldMap" refers to fieldmap in the form of a single NIfTI file 
+# with 2 volumes in it (volume-1: the Fieldmap in Hertz; volume-2: the magnitude image). 
+# Note: dcm2niix (pre-v1.0.20210410) use to convert the GEHC B0Maps as a 2 volumes NIFTI file. 
```
```suggestion
# Note: dcm2niix (pre-v1.0.20210410) used to convert the GEHC B0Maps in this format (a 2 volumes NIFTI file). 
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -9,9 +9,23 @@
+# Note: dcm2niix (pre-v1.0.20210410) use to convert the GEHC B0Maps as a 2 volumes NIFTI file. 
+
+GE_HEALTHCARE_METHOD_OPT="GEHealthCareFieldMap"
+# "GEHealthCareFieldMap" refers to fieldmap in the form of 2 NIfTI files.
```
```suggestion
# "GEHealthCareFieldMap" refers to fieldmap in the form of 2 separate NIfTI files.
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -9,9 +9,23 @@
+
+GE_HEALTHCARE_METHOD_OPT="GEHealthCareFieldMap"
+# "GEHealthCareFieldMap" refers to fieldmap in the form of 2 NIfTI files.
+# One file withe the Fieldmap in Hertz and the magnitude image). 
```
```suggestion
# One file with the Fieldmap in Hz and another file with the magnitude image). 
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -9,9 +9,23 @@
+GE_HEALTHCARE_METHOD_OPT="GEHealthCareFieldMap"
+# "GEHealthCareFieldMap" refers to fieldmap in the form of 2 NIfTI files.
+# One file withe the Fieldmap in Hertz and the magnitude image). 
+# Note: dcm2niix (v1.0.20210410 and later) convert the GEHC B0Maps as 2 NIFTI files
```
```suggestion
# Note: dcm2niix (v1.0.20210410 and later) convert the GEHC B0Maps as 2 separate NIFTI files
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -9,9 +9,23 @@
+# "GEHealthCareFieldMap" refers to fieldmap in the form of 2 NIfTI files.
+# One file withe the Fieldmap in Hertz and the magnitude image). 
+# Note: dcm2niix (v1.0.20210410 and later) convert the GEHC B0Maps as 2 NIFTI files
+# using the suffix _fieldmaphz for the fieldmap in Hertz and no suffix for the magnitude. 
```
```suggestion
# using the suffix '_fieldmaphz' for the fieldmap in Hz and no suffix for the magnitude image. 
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -24,15 +38,17 @@ fi
+source "$HCPPIPEDIR/global/scripts/fsl_version.shlib"          # FSL-version checks functions
 
-opts_SetScriptDescription "Script for generating a fieldmap suitable for FSL from Siemens Gradient Echo field map, and also do gradient non-linearity distortion correction of these"
+opts_SetScriptDescription "Script for generating a fieldmap suitable for FSL from Gradient Echo field map, and also do gradient non-linearity distortion correction of these"
```
```suggestion
opts_SetScriptDescription "Script for generating a fieldmap suitable for FSL from a dual-echo Gradient Echo field map acquisition, and also do gradient non-linearity distortion correction of these"
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -24,15 +38,17 @@ fi
         '${GENERAL_ELECTRIC_METHOD_OPT}'
-             use General Electric specific Gradient Echo Field Maps for SDC
-
+             use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --GEB0InputName argument).
```
```suggestion
             use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (field map in Hz and magnitude image in a single NIfTI file, via --GEB0InputName argument).
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -24,15 +38,17 @@ fi
-
+             use GE HealthCare Legacy specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in a single NIfTI file via --GEB0InputName argument).
+        '${GE_HEALTHCARE_METHOD_OPT}'
+             use GE HealthCare specific Gradient Echo Field Maps for SDC (fieldmap and magnitude in two separate NIfTI files).
```
```suggestion
             use GE HealthCare specific Gradient Echo Field Maps for SDC (field map in Hz and magnitude image in two separate NIfTI files).
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -42,14 +58,15 @@ opts_AddMandatory '--ofmapmagbrain' 'MagnitudeBrainOutput' 'image' "output disto
 
-#Optional Arguments
-opts_AddOptional '--echodiff' 'DeltaTE' 'number (milliseconds)' "echo time difference for Siemens and Philips fieldmap images (in milliseconds)"
+# options --echodiff is now mandatory - used by all "--method" (and all manufacturer). 
```
```suggestion
# options --echodiff is now mandatory - used by all "--method" (and necessary for processing of all vendors). 
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -42,14 +58,15 @@ opts_AddMandatory '--ofmapmagbrain' 'MagnitudeBrainOutput' 'image' "output disto
 
-opts_AddOptional '--fmap' 'GEB0InputName' 'number (degrees)' "input General Electric fieldmap with fieldmap in deg and magnitude image"
+# Optional Arguments
+opts_AddOptional '--fmap' 'GEB0InputName' 'image (hertz and magnitude)' "input GE HealthCare Legacy fieldmap with fieldmap in hertz and magnitude image"
```
```suggestion
opts_AddOptional '--fmap' 'GEB0InputName' 'image (hertz and magnitude)' "GE HealthCare Legacy fieldmap with field map in Hz and magnitude image included as two volumes in a single file"
```

mharms replied:
Also, should we perhaps change the forward facing name of this flag to `--fmapgelegacy` (which including `--fmap` as an additional recognized option for backwards compatibility)?

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -42,14 +58,15 @@ opts_AddMandatory '--ofmapmagbrain' 'MagnitudeBrainOutput' 'image' "output disto
+opts_AddOptional '--fmap' 'GEB0InputName' 'image (hertz and magnitude)' "input GE HealthCare Legacy fieldmap with fieldmap in hertz and magnitude image"
 
-opts_AddOptional '--fmapphase' 'PhaseInputName' 'number (radians)' "input Siemens/Philips fieldmap phase image - in radians"
+opts_AddOptional '--fmapphase' 'PhaseInputName' 'image (radians or hertz)' "input phase image in radians for Siemens/Philips fieldmap and in Hertz for GE HealthCare fieldmap"
```
```suggestion
opts_AddOptional '--fmapphase' 'PhaseInputName' 'image (radians or Hz)' "phase image in radians for Siemens/Philips fieldmap and in Hertz for GE HealthCare fieldmap"
```

mharms commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -42,14 +58,15 @@ opts_AddMandatory '--ofmapmagbrain' 'MagnitudeBrainOutput' 'image' "output disto
+opts_AddOptional '--fmapphase' 'PhaseInputName' 'image (radians or hertz)' "input phase image in radians for Siemens/Philips fieldmap and in Hertz for GE HealthCare fieldmap"
 
-opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'image' "input Siemens/Philips fieldmap magnitude image - can be a 4D containing more than one"
+opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'image' "input Siemens/Philips/GE HealthCare fieldmap magnitude image - can be a 4D containing more than one"
```
```suggestion
opts_AddOptional '--fmapmag' 'MagnitudeInputName' 'image' "Siemens/Philips/GE HealthCare fieldmap magnitude image - can be a 4D containing more than one"
```

mharms commented on global/scripts/fsl_version.shlib:
```diff
@@ -1,3 +1,9 @@
+#!/bin/echo This script should not be run directly:
+
+# Max number of dots in FSL-version strings.
+# We assume the FSL is maximum 4 digits separated 
```
```suggestion
# We assume the FSL version is a maximum of 4 digits separated 
```

mharms commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -32,11 +32,17 @@ source "${HCPPIPEDIR}/global/scripts/fsl_version.shlib"          # Functions for
 SPIN_ECHO_METHOD_OPT="TOPUP"
 NONE_METHOD_OPT="NONE"
 
+# Minimum required FSL version for GENERAL_ELECTRIC_METHOD_OPT and GE_HEALTHCARE_METHOD_OPT
```
Do we need the duplication of checking FSL version for GE where, when, from what I can tell, the same version checks are already built into `FieldMapPreprocessingAll.sh`?

mharms replied:
It also means there are two places in the code making an identical check, and we'd have to remember to change both in any future changes.  @coalsont : How do you want to handle this sort of thing?

mharms reviewed (commented):
I've suggested a bunch of changes to some of the comments and usage statements.  And raised some questions about renaming some of the variables.

brice82 commented:
Few more commits on their way to address the comments of @mharms

brice82 commented:
The latest commits should address the comments from @mharms. 
In Summary: 
- GENERAL_ELECTRIC_METHOD_OPT has been renamed GE_HEALTHCARE_LEGACY_METHOD_OPT
- Removed the TE / DeltaTE confusion in PreFreeSurferPipelineBatch.sh (it is now DeltaTE)
- option --fmapgeneralelectric is now just --fmap (only used in the GE_HEALTHCARE_LEGACY_METHOD_OPT) and all other field map methods uses --fmapmag and --fmapphase 
- copy a number of Mike's corrections from GenericfMRIVolumeProcessingPipelineBatch.sh to GenericfMRIVolumeProcessingPipelineBatch.7T.sh

I left the duplication of checking FSL version for GEHC in GenericfMRIVolumeProcessingPipeline.sh. 
The idea is to throw the error before starting any processing, right at the beginning.
Let me know if you want this FSL version check removed. 

Thanks @mharms for the numerous typos correction and documentation clarification.

coalsont commented:
Yes, erroring early is better, so having the check in two places is fine.

I think the `--fmap` flag was simply "use this if the magnitude and phase are concatenated in a single file", I actually don't think it is correct to rename it based on the mere fact that only one method currently uses that format.  Let's not be excessively specific in option names such that it would make them awkward to use them for a similar purpose with a different brand.  If we wanted to be more specific in a brand-agnostic way, we could call it `--fmapcombined` or similar.

mharms commented:
In that case, I'd favor `--fmapcombined`

mharms commented:
More precisely, make `--fmapcombined` the forward facing flag, but I'm fine with continuing to support `--fmap` as an alternative, given that the parsing library makes that easy to do.

coalsont commented on global/scripts/FieldMapPreprocessingAll.sh:
```diff
@@ -42,14 +58,15 @@ opts_AddMandatory '--ofmapmagbrain' 'MagnitudeBrainOutput' 'image' "output disto
 
-opts_AddOptional '--fmap' 'GEB0InputName' 'number (degrees)' "input General Electric fieldmap with fieldmap in deg and magnitude image"
+# Optional Arguments
+opts_AddOptional '--fmapgelegacy' 'GEB0InputName' 'image (hertz and magnitude)' "GE HealthCare Legacy fieldmap with field map in Hz and magnitude image included as two volumes in a single file"
```
```suggestion
opts_AddOptional '--fmapcombined' 'GEB0InputName' 'image (hertz and magnitude)' "GE HealthCare Legacy fieldmap with field map in Hz and magnitude image included as two volumes in a single file" '' '--fmap'
```

This is how to support a (hidden) legacy flag with our option parsing library.  The variable name could also be made more generic, if you feel like it.

brice82 replied:
integrated in commit [755e64c](https://github.com/Washington-University/HCPpipelines/pull/285/commits/755e64c33bde815fd2aa16f7cd62cb3af79f8c91)

brice82 commented:
To finalize this PR and address the remaining comments from @mharms and @coalsont, I did:
- Use `--fmapcombined` as suggested (and `--fmap` as hidden flag). So there is no more compagny names in the command-line. 
- Moved the definition `GEHEALTHCARE_MINIMUM_FSL_VERSION` in `global/scripts/fsl_version.shlib` as a global constant (now it is defined only once) and added appropriate comments. This should ease the maintenance. 

I feel like we should be good to go now. 
Let me know.

coalsont commented on fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh:
```diff
@@ -315,10 +291,37 @@ case "$DistortionCorrection" in
-			log_Err_Abort "--fmapgeneralelectric must be specified with --dcmethod=${DistortionCorrection}"
+			log_Err_Abort "--fmapcombined must be specified with --dcmethod=${DistortionCorrection}"
+		fi
+    if [ -z ${deltaTE} ]; then
+      log_Err_Abort "--echodiff must be specified with --dcmethod=${DistortionCorrection}"
+    fi
+    # Check that FSL is at least the minimum required FSL version, abort if needed (and log FSL-version)
+    # This FSL version check is duplicated in global/scripts/FieldMapPreprocessingAll.sh
+    # The intention is to catch the error as early as possible. 
+    # GEHEALTHCARE_MINIMUM_FSL_VERSION defined in global/scripts/fsl_version.shlib
+    fsl_minimum_required_version_check "$GEHEALTHCARE_MINIMUM_FSL_VERSION" \
+      "For ${DistortionCorrection} method the minimum required FSL version is ${GEHEALTHCARE_MINIMUM_FSL_VERSION}. " 
+		;;
+  
+  ${GE_HEALTHCARE_METHOD_OPT})
+		if [ -z ${MagnitudeInputName} ]; then
+			log_Err_Abort "--fmapmag must be specified with --dcmethod=${DistortionCorrection}"
+		fi
+		if [ -z ${PhaseInputName} ]; then
+			log_Err_Abort "--fmapphase must be specified with --dcmethod=${DistortionCorrection}"
+		fi
+		if [ -z ${deltaTE} ]; then
+			log_Err_Abort "--echodiff must be specified with --dcmethod=${DistortionCorrection}"
 		fi
+    # Check that FSL is at least the minimum required FSL version, abort if needed (and log FSL-version)
+    # This FSL version check is duplicated in global/scripts/FieldMapPreprocessingAll.sh
+    # The intention is to catch the error as early as possible. 
+    # GEHEALTHCARE_MINIMUM_FSL_VERSION defined in global/scripts/fsl_version.shlib
+    fsl_minimum_required_version_check "$GEHEALTHCARE_MINIMUM_FSL_VERSION" \
+      "For ${DistortionCorrection} method the minimum required FSL version is ${GEHEALTHCARE_MINIMUM_FSL_VERSION}. "
```
This has a mix of tabs and 2 spaces indent.  Sorry for not noticing earlier.  Since most of the surrounding stuff is tabs (and because I don't like 2 spaces indent, not distinct enough for me), I think I'd unify on tabs in this case (probably also fix the Philips section, too).

coalsont commented on Examples/Scripts/Historical/7T/GenericfMRIVolumeProcessingPipelineBatch.7T.sh:
```diff
@@ -183,7 +184,7 @@ do
 
 		# Susceptibility distortion correction method (required for accurate processing)
-		# Values: TOPUP, SiemensFieldMap (same as FIELDMAP), GeneralElectricFieldMap
+        # Values: TOPUP, SiemensFieldMap (same as FIELDMAP), GEHealthCareLegacyFieldMap, GEHealthCareFieldMap, PhilipsFieldMap
```
Indentation mismatch.

coalsont replied:
```suggestion
		# Values: TOPUP, SiemensFieldMap (same as FIELDMAP), GEHealthCareLegacyFieldMap, GEHealthCareFieldMap, PhilipsFieldMap
```

coalsont commented on Examples/Scripts/Historical/7T/GenericfMRIVolumeProcessingPipelineBatch.7T.sh:
```diff
@@ -208,20 +209,36 @@ do
 		
-		# Skipping Gradient Distortion Correction
+		# Gradient distortion correction
+        # Set to NONE to skip gradient distortion correction
+        # (These files are considered proprietary and therefore not provided as part of the HCP Pipelines -- contact Siemens to obtain)
+        # GradientDistortionCoeffs="${HCPPIPEDIR_Config}/coeff_SC72C_Skyra.grad"
```
```suggestion
		# Set to NONE to skip gradient distortion correction
		# (These files are considered proprietary and therefore not provided as part of the HCP Pipelines -- contact Siemens to obtain)
		# GradientDistortionCoeffs="${HCPPIPEDIR_Config}/coeff_SC72C_Skyra.grad"
```

mharms commented:
I don't have anything else to add, other than thanks @brice82 for these changes and additions.  This is a great example of a community contribution!

brice82 commented:
I converted all tabs into 4 spaces in `GenericfMRIVolumeProcessingPipelineBatch.7T.sh` and `GenericfMRIVolumeProcessingPipeline.sh` and also fix the indentation, there were plenty of tabs remaining.
Difference should then be whitespace only. 

Thanks a lot @mharms and @coalsont . 

This time the PR should be good to go.

coalsont commented:
That's one way to do it.  I don't think we cared about "git blame" on the batch file, and it looks like a fair amount of fMRIVolume wasn't using tabs, so most of its "git blame" will survive.

Did you test it after changing to the `--fmapcombined` flag?  I think I'm happy now (sorry for not catching all the formatting in the first review).

brice82 commented:
Yes, I rerun all the test scripts (from PreSurfer until fMRIVolume) with --fmapcombined and --fmap but on a single subject. 
Same output. No error. 
Run test script of FieldMapProccessingAll.sh with --fmap / --fmapcombined (with legacy pipeline) and with --fmapphase and --fmapmag (New processing pipeline) on the 7 subjects. 
Identical results. No error. 
I concluded that everything was as expected.

glasserm commented:
Do folks feel this is ready to merge?
